controller/src/main/java/org/jboss/as/controller/CapabilityRegistry.java

```diff
@@ -940,16 +940,33 @@ private static Set<RuntimeRequirementRegistration> findInconsistent(Map<Capabili
 
     private RuntimeCapabilityRegistration getCapabilityRegistration(String capabilityName, CapabilityScope capabilityScope) {
         SatisfactoryCapability satisfactoryCapability = findSatisfactoryCapability(capabilityName, capabilityScope, false);
+
         if (satisfactoryCapability == null) {
             if (forServer) {
-                throw ControllerLogger.MGMT_OP_LOGGER.unknownCapability(capabilityName);
+                if (isInCapability(capabilityName)) {
+                    throw ControllerLogger.MGMT_OP_LOGGER.cannotGetCapability(capabilityName);
+                } else {
+                    throw ControllerLogger.MGMT_OP_LOGGER.missingCapability(capabilityName);
+                }
+            }
+            if (isInCapability(capabilityName)) {
+                throw ControllerLogger.MGMT_OP_LOGGER.cannotGetCapabilityInContext(capabilityScope.getName(), capabilityName);
             } else {
-                throw ControllerLogger.MGMT_OP_LOGGER.unknownCapabilityInContext(capabilityName, capabilityScope.getName());
+                throw ControllerLogger.MGMT_OP_LOGGER.missingCapabilityInContext(capabilityScope.getName(), capabilityName);
             }
         }
         return capabilities.get(satisfactoryCapability.singleCapability);
     }
 
+    private boolean isInCapability(String capabilityName) {
+        for (CapabilityId key: capabilities.keySet()) {
+            if (capabilities.get(key).getCapabilityName().equals(capabilityName)) {
```

Comment:
This should just be 

if (key.getCapabilityName().equals(capabilityName)) {

It's the same thing.

Note though that if it were different, you should either iterate over capabilities.values() or capabilities.entrySet(). Getting a key set and then calling get is less efficient.